### PR TITLE
Adds a data generation library for analytics data.

### PR DESCRIPTION
**Pull request by danielyule**
This library tries to match some important features of analytics data,
including scoping, time series and aggregating.

### CODE REVIEW

File: src/data_config.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
+import { Category, Measure, Scope, MeasureType } from "./datagen/types";
+
+
+
+
+
```
Comment (Awk34) [inline src/data_config.ts:6]:
```suggestion

```

File: src/data_config.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+  }, {
+    name: 'eventCount',
+    scope: Scope.EVENT,
+    type: MeasureType.SUM,
+  }];
+
+  
+
+  
+
```
Comment (Awk34) [inline src/data_config.ts:64]:
```suggestion

```

File: src/datagen/cube.ts [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+          if (aCategory > bCategory) {
+            return 1;
+          }
+          return 0;
+        };
+      }
+      const measureIndex = measureNames.findIndex(
+        (measure) => measure === sortConcept
+      );
+      if (measureIndex >= 0) {
```
Comment (Awk34) [inline src/datagen/cube.ts:227]:
```suggestion
      if (measureNames.includes(sortConcept)) {
```

File: src/datagen/filters.ts [hunk 4]
```diff
@@ -0,0 +1,14 @@
+import { Category, Row, Filter } from "./types";
+
+const millisecondsPerDay = 86400000;
```
Comment (Awk34) [inline src/datagen/filters.ts:3]: I think it'd be fine to add in the actual multiplications here. A minification transplier would pre-compute it for production code.

File: src/datagen/filters.ts [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+
+export function betweenDates(startDate: Date, endDate: Date): Filter {
+    return (categories: Category[]) => {
+        const nThDayIndex = categories.findIndex((category) => category.name == 'nthDay' );
+        const startIndex = Math.round((Date.now() - startDate.getTime())/ millisecondsPerDay);
```
Comment (Awk34) [inline src/datagen/filters.ts:8]:
```suggestion
        const startIndex = Math.round((Date.now() - startDate.getTime()) / millisecondsPerDay);
```

File: src/datagen/filters.ts [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+        const endIndex = Math.round((Date.now() - endDate.getTime())/ millisecondsPerDay);
```
Comment (Awk34) [inline src/datagen/filters.ts:9]:
```suggestion
        const endIndex = Math.round((Date.now() - endDate.getTime()) / millisecondsPerDay);
```

File: src/datagen/cube.ts [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+ *
+ * This cube is saying, for example, that female Linux users coming from
+ * Search stayed on the page for an average of 5 seconds.
+ *
+ * The cube can be queried for specific catgories and measures.  This
+ * will combine together the values for all the other categories and give
+ * a breakdown based on the requested categories.  For more information, see
+ * the documentation for the `getDataFor()` method.
+ */
+export class Cube {
```
Comment (Awk34) [inline src/datagen/cube.ts:48]: maybe DataCube?

File: src/datagen/generation.ts [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+    );
+    return category.values.map((value) => value.weight / total);
+  });
+}
+
+function binarySearch(
+  arr: number[],
+  probe: number,
+  start: number,
+  end: number
```
Comment (Awk34) [inline src/datagen/generation.ts:138]:
```suggestion
  start = 0,
  end = arr.length
```

File: src/data_config.ts (hunk 1 above)
Comment (danielyule) [reply src/data_config.ts:6]: No idea why I left those in.

File: src/datagen/cube.ts (hunk 3 above)
Comment (danielyule) [reply src/datagen/cube.ts:227]: Good suggestion.

File: src/datagen/cube.ts (hunk 7 above)
Comment (danielyule) [reply src/datagen/cube.ts:48]: Good suggestiion

File: src/datagen/filters.ts (hunk 4 above)
Comment (danielyule) [reply src/datagen/filters.ts:3]: I had actually googled this number directly.  Changed to computation.

**Pull request merged**